File: weather-demo-app-apple/Weather/UI/Forecast/ForecastViewModel.swift
//
//  ForecastViewModel.swift
//  Weather
//
//  Created by Onur on 4/11/26.
//

import Foundation
import Combine

@MainActor
final class ForecastViewModel: ObservableObject {
    @Published private(set) var state: ForecastViewState = .idle

    private let repository: WeatherDataSource
    private var lastLoadedCity: String?
    private var loadTask: Task<Void, Never>?

    init(repository: WeatherDataSource) {
        self.repository = repository
    }

    func load(city: String, force: Bool = false) async {
        let targetCity = city.trimmingCharacters(in: .whitespacesAndNewlines)

        if !force,
           lastLoadedCity == targetCity,
           case .success = state {
            return
        }

        loadTask?.cancel()

        loadTask = Task {
            state = .loading

            do {
                let days = try await repository.getForecast(city: targetCity)
                self.lastLoadedCity = targetCity
                self.state = .success(days)
            } catch let error as RepositoryError {
                self.state = .error(error.message)
            } catch {
                self.state = .error(AppConstants.genericErrorMessage)
            }
        }

        await loadTask?.value
    }

    func retry(city: String) async {
        await load(city: city, force: true)
    }
}

File: weather-demo-app-android/app/src/main/java/com/demo/weather/ui/forecast/ForecastViewModel.kt
package com.demo.weather.ui.forecast

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.demo.weather.app.AppConstants
import com.demo.weather.data.WeatherDataSource
import com.demo.weather.domain.model.RepositoryError
import com.demo.weather.domain.model.toMessage
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class ForecastViewModel(
    private val repository: WeatherDataSource
) : ViewModel() {

    private val _state = MutableStateFlow<ForecastViewState>(ForecastViewState.Idle)
    val state: StateFlow<ForecastViewState> = _state.asStateFlow()

    private var lastLoadedCity: String? = null
    private var loadJob: Job? = null

    fun load(city: String, force: Boolean = false) {
        val targetCity = city.trim()

        if (!force &&
            lastLoadedCity == targetCity &&
            _state.value is ForecastViewState.Success
        ) {
            return
        }

        loadJob?.cancel()
        loadJob = viewModelScope.launch {
            _state.value = ForecastViewState.Loading

            try {
                val days = repository.getForecast(targetCity)
                lastLoadedCity = targetCity
                _state.value = ForecastViewState.Success(days)
            } catch (error: RepositoryError) {
                _state.value = ForecastViewState.Error(error.toMessage())
            } catch (_: Throwable) {
                _state.value = ForecastViewState.Error(AppConstants.GENERIC_ERROR_MESSAGE)
            }
        }
    }

    fun retry(city: String) {
        load(city = city, force = true)
    }
}